src/crimson/osd/pg.cc

```diff
@@ -517,71 +535,98 @@ seastar::future<Ref<MOSDOpReply>> PG::do_osd_ops(
                                                    : m->get_hobj();
   auto ox =
     std::make_unique<OpsExecuter>(obc, *this/* as const& */, m);
-  return crimson::do_for_each(
-    m->ops, [this, obc, m, ox = ox.get()](OSDOp& osd_op) {
-    logger().debug(
-      "do_osd_ops: {} - object {} - handling op {}",
-      *m,
-      obc->obs.oi.soid,
-      ceph_osd_op_name(osd_op.op.op));
-    return ox->execute_osd_op(osd_op);
-  }).safe_then([this, obc, m, ox = std::move(ox)] {
-    logger().debug(
-      "do_osd_ops: {} - object {} all operations successful",
-      *m,
-      obc->obs.oi.soid);
-    return std::move(*ox).submit_changes(
-      [this, m] (auto&& txn, auto&& obc) -> osd_op_errorator::future<> {
-        // XXX: the entire lambda could be scheduled conditionally. ::if_then()?
-        if (txn.empty()) {
+
+  OSDOpParams osd_op_p(m);
```

Comment:
How about making this a part of `OpsExecuter`? That way you could reuse the big `switch (osd_op.op.op)` we already have in `execute_osd_op()` for `user_modify`. There is `do_write_op()` logic which might be useful.
Also, IIRC `seastar::do_with()` imposes extra `memcpy` we squeezed with `auto ox = std::make_unique<OpsExecuter>`. It's rather small issue but my general impression was it's better to avoid `do_with()` in performance-critical areas.

Reply:
Ok, I'll try your way:-) Thanks:-)